### PR TITLE
New kde-neon extension for Qt6 / KF6

### PR DESCRIPTION
**Pull request by ScarlettGatelyMoore**
- [x ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x ] Have you successfully run `tox run -m lint`?
- [ x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

Tests pass locally.
-----

### CODE REVIEW

**Review by mr-cal — APPROVED** [review]
Thanks!  Spread tests are failing due to https://forum.snapcraft.io/t/kde-global-auto-connect-request/39535

File: extensions/desktop/kde-neon-6/locale-gen [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+use strict;
+
+sub uniq
+{
+    my %seen;
+    grep !$seen{$_}++, @_;
+}
+
+# TODO support for KDE's in-app language switch feature.
```
Comment (lengau) [inline extensions/desktop/kde-neon-6/locale-gen:11]: (Not a blocker, just a question) Roughly what will it take to support that feature?

File: snapcraft/extensions/kde_neon_6.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2022 Canonical Ltd.
+#           2023 Scarlett Moore <sgmoore@kde.org>
```
Comment (lengau) [inline snapcraft/extensions/kde_neon_6.py:4]:
```suggestion
#           2023-2024 Scarlett Moore <sgmoore@kde.org>
```

File: snapcraft/extensions/kde_neon_6.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+import functools
+import re
+from typing import Any, Dict, List, Optional, Tuple
+
+from overrides import overrides
+
+from .extension import Extension, get_extensions_data_dir, prepend_to_env
+
+_QT6_SDK_SNAP = {"core22": "kde-qt6-core22-sdk"}
+_KF6_SDK_SNAP = {"core22": "kf6-core22-sdk"}
```
Comment (lengau) [inline snapcraft/extensions/kde_neon_6.py:29]: I gather we're safe on backwards compatibility of the ABIs for both? (No need for minor version specifications?)

File: tests/unit/extensions/test_kde_neon_6.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2022 Canonical Ltd.
```
Comment (lengau) [inline tests/unit/extensions/test_kde_neon_6.py:3]:
```suggestion
# Copyright 2022 Canonical Ltd.
# Copyright 2024 Scarlett Moore <sgmoore@kde.org>
```

**Review by lengau — APPROVED** [review]
Looks good, thanks!

- One related question (ABI stability)
- One curiosity question
- Two minor change requests to satisfy lawyers

File: snapcraft/extensions/kde_neon_6.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        return {
+            "build-environment": [
+                {
+                    "PATH": prepend_to_env(
+                        "PATH",
+                        [
+                            f"/snap/{qt6_sdk_snap}/current/"
+                            f"usr/bin:/snap/{kf6_sdk_snap}/"
+                            f"current/usr/bin"
+                        ],
```
Comment (lengau) [inline snapcraft/extensions/kde_neon_6.py:219]:
```suggestion
                        [
                            f"/snap/{qt6_sdk_snap}/current/usr/bin",
                            f"/snap/{kf6_sdk_snap}/current/usr/bin",
                        ],
```

This looks more readable to me

File: snapcraft/extensions/kde_neon_6.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                    "XDG_DATA_DIRS": prepend_to_env(
+                        "XDG_DATA_DIRS",
+                        [
+                            f"$CRAFT_STAGE/usr/share:"
+                            f"/snap/{qt6_sdk_snap}"
+                            f"/current/usr/share:"
+                            f"/snap/{kf6_sdk_snap}"
+                            f"/current/usr/share",
+                            "/usr/share",
+                        ],
```
Comment (lengau) [inline snapcraft/extensions/kde_neon_6.py:232]:
```suggestion
                        [
                            f"$CRAFT_STAGE/usr/share",
                            f"/snap/{qt6_sdk_snap}/current/usr/share",
                            f"/snap/{kf6_sdk_snap}/current/usr/share",
                            "/usr/share",
                        ],
```

File: snapcraft/extensions/kde_neon_6.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+                    ),
+                },
+                {
+                    "SNAPCRAFT_CMAKE_ARGS": prepend_to_env(
+                        "SNAPCRAFT_CMAKE_ARGS",
+                        [
+                            cmake_args,
+                        ],
+                    ),
+                },
```
Comment (lengau) [inline snapcraft/extensions/kde_neon_6.py:242]: Looking at the [ark snap](https://invent.kde.org/scarlettmoore/ark/-/blob/master/snapcraft.yaml?ref_type=heads), would these additions make sense?

They work for me locally and still let me build an apparently-functional snap of ark, but I'm not sure what downsides there may be:

```suggestion
                },
                {
                    "XDG_CONFIG_HOME": prepend_to_env(
                        "XDG_CONFIG_HOME",
                        [
                            "$CRAFT_STAGE/etc/xdg",
                            f"/snap/{qt6_sdk_snap}/current/etc/xdg",
                            f"/snap/{kf6_sdk_snap}/current/etc/xdg",
                            "/etc/xdg",
                        ],
                    ),
                },
                {
                    "LD_LIBRARY_PATH": prepend_to_env(
                        "LD_LIBRARY_PATH",
                        [
                            f"/snap/{qt6_sdk_snap}/current/usr/lib/${{CRAFT_ARCH_TRIPLET_BUILD_FOR}}",
                            f"/snap/{kf6_sdk_snap}/current/usr/lib/${{CRAFT_ARCH_TRIPLET_BUILD_FOR}}",
                            f"/snap/{qt6_sdk_snap}/current/usr/lib",
                            f"/snap/{kf6_sdk_snap}/current/usr/lib",
                            "${CRAFT_STAGE}/usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}"
                            "${CRAFT_STAGE}/usr/lib",
                            "${CRAFT_STAGE}/lib/",
                        ],
                    ),
                },
                {
                    "CMAKE_PREFIX_PATH": prepend_to_env(
                        "CMAKE_PREFIX_PATH",
                        [
                            "$CRAFT_STAGE",
                            f"/snap/{qt6_sdk_snap}/current",
                            f"/snap/{kf6_sdk_snap}/current",
                            "/usr",
                        ],
                        separator=";",
                    ),
                },
                {
                    "CMAKE_FIND_ROOT_PATH": prepend_to_env(
                        "CMAKE_FIND_ROOT_PATH",
                        [
                            "$CRAFT_STAGE",
                            f"/snap/{qt6_sdk_snap}/current",
                            f"/snap/{kf6_sdk_snap}/current",
                            "/usr",
                        ],
                        separator=";",
                    ),
                },
```

File: snapcraft/extensions/kde_neon_6.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        if self.kde_snaps.qt6_builtin:
+            return {
+                "kde-neon-6/sdk": {
+                    "source": str(source),
+                    "plugin": "make",
+                    "make-parameters": [f"PLATFORM_PLUG={self.kde_snaps.content_kf6}"],
+                    "build-snaps": [
+                        self.kde_snaps.qt6_sdk_snap,
+                        self.kde_snaps.kf6_sdk_snap,
+                    ],
```
Comment (lengau) [inline snapcraft/extensions/kde_neon_6.py:262]:
```suggestion
                    ],
                    "build-packages": [
                        "gettext",
                        "libxml2-utils",
                        "docbook-xml",
                        "docbook-xsl",
                        "libglx-dev",
                        "libgl1-dev",
                        "libglvnd-dev",
                    ],
```

These seem to be bulid requirements for pretty much everything - any reason not to include them always?

File: extensions/desktop/kde-neon-6/locale-gen (hunk 1 above)
Comment (ScarlettGatelyMoore) [reply extensions/desktop/kde-neon-6/locale-gen:11]: That isn't my TO-DO. I wouldn't even know where to start :(

File: snapcraft/extensions/kde_neon_6.py (hunk 3 above)
Comment (ScarlettGatelyMoore) [reply snapcraft/extensions/kde_neon_6.py:29]: This is fine. Previously we were using packages and had to ensure the package versions were aligned. We are now building all the parts from source. If there are incompatible versions it is a snap problem that needs attention.

Comment (lengau) [reply snapcraft/extensions/kde_neon_6.py:29]: Awesome, thanks!

Comment (sergiusens) [reply snapcraft/extensions/kde_neon_6.py:29]: this is great news!

File: extensions/desktop/kde-neon-6/locale-gen (hunk 1 above)
Comment (sergiusens) [reply extensions/desktop/kde-neon-6/locale-gen:11]: this was Harald IIRC, so that is an entry point

File: snapcraft/extensions/kde_neon_6.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    \b
+    - desktop (https://snapcraft.io/docs/desktop-interface)
+    - desktop-legacy (https://snapcraft.io/docs/desktop-legacy-interface)
+    - opengl (https://snapcraft.io/docs/opengl-interface)
+    - wayland (https://snapcraft.io/docs/wayland-interface)
+    - x11 (https://snapcraft.io/docs/x11-interface
+    - audio-playback
+    - unity7
+    - network
+    - network-bind
```
Comment (sergiusens) [inline snapcraft/extensions/kde_neon_6.py:69]:
```suggestion
    - x11 (https://snapcraft.io/docs/x11-interface)
    - audio-playback (https://snapcraft.io/docs/audio-playback-interface)
    - unity7 (https://snapcraft.io/docs/unity7-interface)
    - network https://snapcraft.io/docs/network-interface)
    - network-bind (https://snapcraft.io/docs/network-bind-interface)
```

File: extensions/desktop/kde-neon-6/launcher-specific [hunk 10]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
```
Comment (sergiusens) [inline extensions/desktop/kde-neon-6/launcher-specific:1]: for the gnome extension we went with putting these scripts in the content snap itself, to make changes and updates easier. I understand that can be a lot more work, so if it is too much effort now, I would at least encourage it for the core24 version of these snaps.

Not going to block on this

**Review by sergiusens — APPROVED** [review]
great work!

File: extensions/desktop/kde-neon-6/launcher-specific (hunk 10 above)
Comment (ScarlettGatelyMoore) [reply extensions/desktop/kde-neon-6/launcher-specific:1]: Will do for core24, thanks

File: snapcraft/extensions/kde_neon_6.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+            qt6_builtin = True
+
+        matcher = re.compile(r"kf6-" + base + r"-sdk.*")
+        kf6_sdk_snap_candidates = [s for s in build_snaps if matcher.match(s)]
+        if kf6_sdk_snap_candidates:
+            kf6_sdk_snap = kf6_sdk_snap_candidates[0].split("/")[0]
+            kf6_builtin = False
+        else:
+            kf6_builtin = True
+        # The same except the trailing -sd
```
Comment (Saviq) [inline snapcraft/extensions/kde_neon_6.py:130]:
```suggestion
        # The same except the trailing -sdk
```

**Comment by sergiusens** [comment]
seems like the connection is not happening in the test:

     + '[' 'ERROR: not connected to the kf6-core22 content interface.' = 'hello world' ']'

do we wait for this or manually connect for now? I am ok with either, but I am intending to cut a release tomorrow

**Comment by ScarlettGatelyMoore** [comment]
> seems like the connection is not happening in the test:
> 
> ```
>  + '[' 'ERROR: not connected to the kf6-core22 content interface.' = 'hello world' ']'
> ```
> 
> do we wait for this or manually connect for now? I am ok with either, but I am intending to cut a release tomorrow

I am good with manually connect if it's not to late.

**Comment by ScarlettGatelyMoore** [comment]
I just had to add fix 2b439f9e75c1e05e1aa5219a455c8239c3e79f9a to our $HOME ref: https://forum.snapcraft.io/t/setting-home-in-snap-to-user-home-seems-to-cause-all-empty-user-dirs-to-become-broken-symlinks/39463/5

**Comment by sergiusens** [comment]
@ScarlettGatelyMoore manual connect is good if you can do it now

**Pull request merged**